hw/top_earlgrey/syn/asic.constraints.sdc

```diff
@@ -215,30 +215,109 @@ set_output_delay ${SPI_HOST_OUT_DEL} [get_ports SPI_HOST_D1]   -clock SPI_HOST_C
 set_output_delay ${SPI_HOST_OUT_DEL} [get_ports SPI_HOST_D2]   -clock SPI_HOST_CLK
 set_output_delay ${SPI_HOST_OUT_DEL} [get_ports SPI_HOST_D3]   -clock SPI_HOST_CLK
 
+#####################################
+# SPI DEV clock Passthru Operation  #
+#####################################
+# Passthrough target freq is 33MHz. Using 40MHz to over-constrain
+#
+# The constraints below take the following approach:
+# Define incoming passthrough clock on the SPI_DEV_CLK pin and relate all the inputs to it.
+# Deinfe also output delays since all pins are bidirectiona.
+# Define outgoing passthrough clock on the SPI_HOST_CLK pin but make sure it is a generated versio
+# of the incoming passthrough clock, relate the host side pins to this clock in both input/output
+# directions.
+
+set SPI_DEV_PASSTHRU_CK 25.0
+create_clock -name SPI_DEV_PASSTHRU_CLK -period ${SPI_DEV_PASSTHRU_CK} [get_ports ${SPI_DEV_CLK_PIN}] -add
+set_clock_uncertainty ${SETUP_CLOCK_UNCERTAINTY} [get_clocks SPI_DEV_PASSTHRU_CLK]
+
+# clocks used by spi device internally
+# Unlike spi-dev above, here the "incoming" clock is treated lie the "launch", while the inverted
+# is treated like "capture".  The other way would work fine as well, but several other adjustments
+# would need to be made the rest of the constraints.
+create_generated_clock -name SPI_DEV_PASSTHRU_IN_CLK -source SPI_DEV_CLK -divide_by 1 \
+    -invert [get_pins top_earlgrey/u_spi_device/u_clk_spi_in_buf/clk_o] -add -master_clock SPI_DEV_PASSTHRU_CLK
+create_generated_clock -name SPI_DEV_PASSTHRU_OUT_CLK -source SPI_DEV_CLK -divide_by 1 \
+    [get_pins top_earlgrey/u_spi_device/u_clk_spi_out_buf/clk_o] -add -master_clock SPI_DEV_PASSTHRU_CLK
+
+# clocks accounting for propagation delay to the other side
+create_generated_clock -name SPI_HOST_PASSTHRU_CLK -source SPI_DEV_CLK -master_clock SPI_DEV_PASSTHRU_CLK -divide_by 1 \
+    [get_ports SPI_HOST_CLK] -add
+
+# The propagated properties are needed to ensure the passthrough clocks assume the passthrough delay.
+set_propagated_clock [get_clock SPI_DEV_PASSTHRU_CLK]
+set_propagated_clock [get_clock SPI_HOST_PASSTHRU_CLK]
+
+# delays below are nominal since target frequency is already over constrained
+set PCB_DEL 1
+set HOST_SETUP_DEL 4
+set HOST_OUT_DELY 3
+set STORAGE_SETUP_DEL 3
+set STORAGE_OUT_DEL 7
+set HALF_CYCLE [expr ${SPI_DEV_PASSTHRU_CK} / 2]
+
+# These are delays facing the host
+set SPI_DEV_PASSTHRU_HOST_IN_DEL [expr 2*${PCB_DEL} + ${HOST_OUT_DELY}]
+set SPI_DEV_PASSTHRU_HOST_OUT_DEL [expr {${PCB_DEL} + ${HOST_SETUP_DEL}}]
+
+# for transactions passing from storage into the device, they are going to be sampled
+# by the host at "full cycle" boundaries
+set SPI_DEV_PASSTHRU_STORAGE_IN_DEL [expr {2*${PCB_DEL} + ${STORAGE_OUT_DEL}}]
+
+# for transactions passing through the device to the storage device, the commands
+# are captured on half cycle boundaries
+set SPI_DEV_PASSTHRU_STORAGE_OUT_DEL [expr ${PCB_DEL} + ${STORAGE_SETUP_DEL} + ${HALF_CYCLE}]
```

Comment:
instead of using `HALF_CYCLE`, should try `output delay` relative to falling edge.

Reply:
Is the capture logic actually on the falling edge?  You could use -invert on the clock somehow instead if it's capturing on the rising edge but phase-shifted by half a cycle via the inversion.